---
title: "ngsgenotype"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(engine.opts = list(bash = "-l"))
```

# Load libraries and settings:
```{r cars}
library(ShortRead)
library(wesanderson)

primerPath <- "./Primers/primers.fa"
filteredFolder <- "./Filtered_data/"
```

# Load primers:
```{r pressure, echo=FALSE}
ParsePrimers <- function(fastaFile) {
  fastaString <- paste(readLines(fastaFile), collapse = "\n")
  fastaList <- strsplit(fastaString, split = ">")
  fastaList <- fastaList[[1]][fastaList[[1]] != ""]
  newLineSplit <- function(inputLine) {
    sequenceList <- strsplit(inputLine, split = "\n")[[1]]
    nucSequence <- sequenceList[2:length(sequenceList)]
    nucSequence <- paste(nucSequence, collapse = "")
    sequenceList <- c(sequenceList[1], strsplit(nucSequence, "\\.\\.\\."))
    sequenceDF <- data.frame(Primer = c(sequenceList[[2]][1], sequenceList[[2]][2]))
    colnames(sequenceDF) <- sequenceList[[1]]
    return(sequenceDF)
  }
  fastaDFList <- lapply(fastaList, newLineSplit)
  fastaDF <- do.call("cbind", fastaDFList)
  return(fastaDF)
}

primerDF <- ParsePrimers(fastaFile = primerPath)
```

# Detect files:
```{r pressure, echo=FALSE}
NameShortener <- function(name) {
  # name <- substr(name, 1, 20)
  # name <- gsub("_[^_]+$", "", name)
  # name <- gsub("BMK240416-BZ163-ZX01-010005-01", "", name)
  # name <- gsub("_Unknown", "", name)
  name <- gsub(".fastq.gz", "", name)
  return(name)
}

FastqLoader <- function(fileName, pathToData = filteredFolder) {
  #namePattern <- paste0(fileName, "*")
  loadedFile <- ShortRead::readFastq(dirPath = pathToData, pattern = fileName)
  print(paste("Searching for", fileName))
  print(paste("Found", length(list.files(pathToData, pattern = fileName)), "files."))
  return(c(loadedFile, fileName))
}

# Find all files and modify names:
inputFiles <- list.files(filteredFolder, pattern = ".fastq.gz")
inputFiles <- unlist(lapply(inputFiles, NameShortener))
```

# Identify genotypes and plot peaks:
```{r pressure, echo=FALSE}
FragLength <- function(sample, primerF = "AAAAA", primerR = "AAAAA", range = 0:1000) {
  lengthF <- nchar(primerF)
  lengthR <- nchar(primerR)
  fragl <- ShortRead::width(ShortRead::sread(sample))
  lengthTot <- lengthF + lengthR + fragl
  lengthTot <- lengthTot[lengthTot > min(range) & lengthTot < max(range)]
  tabl <- table(lengthTot)
  if(length(tabl) > 8) {
    tabl <- sort(tabl, decreasing = TRUE)[1:8]
  }
  # tabl <- as.table(tabl[tabl > cutoff * sum(tabl)])
  names(attributes(tabl)$dimnames) <- "AlleleLength"
  tabl <- tabl[order(as.numeric(names(tabl)))]
  return(tabl)
}

MostLikely <- function(results) {
  AlleleCandidate1 <- results[which.max(names(results))]
  resultsShort <- results[results != AlleleCandidate1]
  AlleleCandidate2 <- resultsShort[which.max(names(resultsShort))]
  results_sorted <- sort(results)
  max_peak <- results_sorted[length(results_sorted) - 0:1]
  AlleleCandidate3 <- max_peak[1]
  AlleleCandidate4 <- max_peak[2]
  AC <- list(AC1 = AlleleCandidate1[1], AC2 = AlleleCandidate2, 
             AC3 = AlleleCandidate3, AC4 = AlleleCandidate4)
  GT <- c(AC$AC1, AC$AC3)
  names(rev(GT))
}

BarCol <- function(Allelelength, highlight) {
  c1 <- rep(wesanderson::wes_palette("Chevalier1")[3], length(Allelelength))
  c1[which(names(Allelelength) %in% highlight)] <- wesanderson::wes_palette("Chevalier1")[2]
  c1
}

PlotSample <- function(sampleName, sampleTab, sampleML) {
  barplot(sampleTab,
          col = BarCol(sampleTab, highlight = sampleML),
          main = sampleName,
          xlab = paste(sampleML, collapse = "/"))
}

PlotAllSamples <- function(loadedFiles, writeTo, primerInfo, fragRange = 0:1000) {
  pdf(writeTo)
  outputDF <- data.frame(Allele = character(), LengthShort = numeric(), LengthLong = numeric())
  for(i in loadedFiles) {
      dfRow <- nrow(outputDF) + 1
      if(length(i[[1]]) == 0) {
          barplot(c(0, 0),
                  main = unlist(i[2]),
                  xlab = "0/0")
          outputDF[dfRow, 1] <- unlist(i[2])
          outputDF[dfRow, 2:3] <- c(NA,NA)
    } else {
      primerName <- unlist(strsplit(unlist(i[2]), "_"))[1]
      primerFR <- primerInfo[, colnames(primerInfo) == primerName]
      tryCatch(
        {
        tabSample <- FragLength(sample = i[[1]], primerF = primerFR[1], primerR = primerFR[2], range = fragRange)
      mlSample <- MostLikely(tabSample)
      PlotSample(sampleName = unlist(i)[2], sampleTab = tabSample, sampleML = mlSample)
      outputDF[dfRow,1] <- unlist(i[2])
      outputDF[dfRow, 2:3] <- as.numeric(mlSample)
        },
        error=function(e) {
            message('An Error Occurred')
            print(e)
            outputDF[dfRow,1] <- unlist(i[2])
            outputDF[dfRow, 2:3] <- c(NA,NA)
        },
        #if a warning occurs, tell me the warning
        warning=function(w) {
            message('A Warning Occurred')
            print(w)
            return(NA)
        }
      )
    }
  }
  dev.off()
  return(outputDF)
}

RunLocus <- function(locusName, allFiles, filteredFolder, locusCutoffs = c()) {
  print(paste0("Running locus: ", locusName))
  locusFiles <- allFiles[grepl(locusName, inputFiles)]
  loadedFastqFiles <- lapply(locusFiles, FastqLoader, filteredFolder)
  if(locusName %in% names(locusCutoffs)) { # if a specific cutoff is set for this locus
    cutoff = locusCutoffs[locusName][[1]]
  } else { # else use standard liberal settings
    cutoff = c(10, 700)
  }
  gentypRes <- PlotAllSamples(loadedFiles = loadedFastqFiles, writeTo = paste0("Plots/", locusName, ".pdf"), primerInfo = primerDF, fragRange = cutoff)
  rm(loadedFastqFiles) # remove to conserve memory
  gc() # run garbage collection to free memory
  colnames(gentypRes) <- c("Sample", paste0(locusName, "_1"), paste0(locusName, "_2"))
  gentypRes$Sample <- gsub(".+_|BMK.+", "", gentypRes$Sample)
  return(gentypRes)
}

RunAllLoci <- function(inputFiles, filteredFolder, locusCutoffs = c()) {
  lociNames <- unique(gsub("_.+", "", inputFiles)) # remove everything after and including the first underscore
  genotypesList <- lapply(lociNames, RunLocus, inputFiles, filteredFolder, locusCutoffs)
  allGenotypes <- Reduce(function(x, y) merge(x, y, by = "Sample"), genotypesList)
  return(allGenotypes)
}

# Specify if any loci should have specific cutoffs:
locusCutoffs <- c()
locusCutoffs <- rep(list(c(10, 1000)), length(unique(gsub("_.+", "", inputFiles))))
names(locusCutoffs) <- unique(gsub("_.+", "", inputFiles))
# locusCutoffs$`locus name` <- c(10, 1000) # add 'locus name' and its designated cutoffs

# Load all samples and calculate all loci:
allGenotypes <- RunAllLoci(inputFiles, filteredFolder, locusCutoffs)

# Write all genotypes to an excel-sheet:
library(writexl)
write_xlsx(x = allGenotypes, path = "all_735_genotypes.xlsx")
```
