---
title: "ngsgenotype_snp"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Detect and load files:
```{r}
# Load libraries:
library(ShortRead)
library(wesanderson)
source("genotyping_functions.r")

# List files:
snpCsvFiles <- list.files("Genotyped_SNPs/", pattern = ".csv", full.names = TRUE)

# Load csv files into a list of dataframes:
loadedFiles <- lapply(snpCsvFiles, read.table, sep = ',', header = TRUE, check.names = FALSE)

# Row bind all separate dataframes into one:
allGenotypes <- do.call(rbind, loadedFiles)

# Replace NA values:
allGenotypes[is.na(allGenotypes)] <- ''
```

# Filter samples and loci:
```{r}
GenotypeFiltering <- function(inGenotypes, cutOff = 20) { # edge case: if a sample has 0% missing data while also having 21 cases of the exact same allele composition it will also be filtered out.
  filteredGenotypes <- inGenotypes[!grepl("NEGATIVE|POSITIVE", inGenotypes$Sample), ]
  filteredGenotypes <- filteredGenotypes[unlist(lapply(apply(filteredGenotypes[, -c(1)], 1, table), `[[`, 1)) <= cutOff, ] # check how many nothing values
  saveRDS(filteredGenotypes, file = "filteredGenotypes.rds")
  rejectedGenotypes <- inGenotypes[unlist(lapply(apply(inGenotypes[, -c(1)], 1, table), `[[`, 1)) > cutOff, ]
  rejectedGenotypes <- rbind(rejectedGenotypes, inGenotypes[grepl("NEGATIVE|POSITIVE", inGenotypes$Sample), ])
  saveRDS(rejectedGenotypes, file = "rejectedGenotypes.rds")
  outGT <- list()
  outGT$filtered <- filteredGenotypes
  outGT$rejected <- rejectedGenotypes
  return(outGT)
}

# Filter out sex loci:
autoGenotypes <- allGenotypes[, !grepl("Ua-snp-01y|Ua-snp-02y|Ua-snp-03y|Ua-snp-04y|Ua-snp-196-x|Ua-snp-248-x", colnames(allGenotypes))]
saveRDS(autoGenotypes, file = "autoGenotypes.rds")
sexGenotypes <- cbind(Sample = allGenotypes$Sample, allGenotypes[, grepl("Ua-snp-01y|Ua-snp-02y|Ua-snp-03y|Ua-snp-04y|Ua-snp-196-x|Ua-snp-248-x", colnames(allGenotypes))])
saveRDS(sexGenotypes, file = "sexGenotypes.rds")

# Filter out samples with less than 66 working loci:
filterOutput <- GenotypeFiltering(autoGenotypes)
filteredGenotypes <- filterOutput$filtered
rejectedGenotypes <- filterOutput$rejected

# Filter out empty loci:
presentGenotypes <- filteredGenotypes[, !lapply(filteredGenotypes, unique) == '']
```

# Collapse samples to individuals:
```{r}
# Load libraries:
library(poppr)
library(adegenet)

# Load data:
genotypeGenind <- adegenet::df2genind(presentGenotypes[, -c(1)], ploidy = 2, ind.names = presentGenotypes$Sample, sep = "")
genotypeGenind
genotypeGencln <- poppr::as.genclone(genotypeGenind)
genotypeGencln

# Collapse genotypes:
mll(genotypeGencln) <- "original" # reset genotypes
system.time(calculatedDistances <- BearDist(presentGenotypes)) # calculate BearDists (TM)
saveRDS(calculatedDistances, file = "calculatedDistances.rds") # save calculated distances
calculatedDistances # print BearDists (TM)
hist(calculatedDistances) # histogram of BearDists (TM)
threshold <- (ncol(presentGenotypes) - 1) - 67 # make a threshold where 67 loci needs to be similar between samples
mlg.filter(genotypeGencln, distance = calculatedDistances) <- threshold # collapse based on threshold, using BearDists (TM)
genotypeGencln
mll(genotypeGencln)

# Assign genotypes:
gtAndSnps <- as.data.frame(cbind(genotypes = genotypeGencln@mlg@mlg$contracted, genotypeGencln$tab)) # add genotypes to SNP table
uniqueSamples <- rownames(gtAndSnps)[match(unique(gtAndSnps$genotypes), gtAndSnps$genotypes)] # take first sample of each genotype
# Pull those samples from the sample sheets

# Mockup:
# 1. add a column to the "database" with ind-names
# 2. load the "database" (without ind-names) and new samples together
# 3. collapse individuals (will collapse the new ones to an existing genotype)
# 4. check if collapsed genotypes contain a sample with an existing ind-name, if so add new ones to database under that name
# 5. if individuals do not exist, add them to database with new names.
```














