---
title: "ngsgenotype_snp"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Detect and load files:
```{r}
# Load libraries:
library(ShortRead)
library(wesanderson)
source("genotyping_functions.r")

# List files:
filteredFolder <- "./Filtered_data/SNP_filtered/"
inputFiles <- list.files(filteredFolder, pattern = ".fastq.gz")

# Generate cutoffs:
locusCutoffs <- GenerateCutoffs(minReads = 10, minPerc = 20)
# locusCutoffs$`locus name` <- c(10, 10) # add 'locus name' and its designated cutoffs

# Load all samples and calculate all loci:
allGenotypes <- RunAllLoci(inputFiles, filteredFolder, locusCutoffs)

# If genotypes are returned as only one letter it is because of the fact that there are empty reads in the fastq
```

# Collapse samples to individuals:
```{r}
# Load libraries:
library(poppr)
library(adegenet)

# Load data:
genotypeGenind <- adegenet::df2genind(allGenotypes[, -c(1)], ploidy = 2, ind.names = allGenotypes$Sample, sep = "")
genotypeGenind
genotypeGencln <- poppr::as.genclone(genotypeGenind)
genotypeGencln

# Collapse genotypes:
mll(genotypeGencln) <- "original" # reset genotypes
BearDist(allGenotypes) # print BearDists (TM)
hist(BearDist(allGenotypes)) # histogram of BearDists (TM)
threshold <- (ncol(allGenotypes) - 1) - 67 # make a threshold where 67 loci needs to be similar between samples
mlg.filter(genotypeGencln, distance = BearDist(allGenotypes)) <- 48 # threshold # collapse based on threshold, using BearDists (TM)
genotypeGencln
mll(genotypeGencln)

# Assign genotypes:
gtAndSnps <- as.data.frame(cbind(genotypes = genotypeGencln@mlg@mlg$contracted, genotypeGencln$tab)) # add genotypes to SNP table
uniqueSamples <- rownames(gtAndSnps)[match(unique(gtAndSnps$genotypes), gtAndSnps$genotypes)] # take first sample of each genotype
# Pull those samples from the sample sheets

# Mockup:
# 1. add a column to the "database" with ind-names
# 2. load the "database" (without ind-names) and new samples together
# 3. collapse individuals (will collapse the new ones to an existing genotype)
# 4. check if collapsed genotypes contain a sample with an existing ind-name, if so add new ones to database under that name
# 5. if individuals do not exist, add them to database with new names.
```
