---
title: "ngsgenotype"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(engine.opts = list(bash = "-l"))
```


# Load libraries and settings:
```{r cars}
library(ShortRead)
library(wesanderson)

primerPath <- "./Primers/primers.fa"
filteredFolder <- "./Filtered_data/"

filteredFolder <- "~/Documents/bearNGSGenotype/finalanalys1/outfor/test2"

```

# Load primers:
```{r pressure, echo=FALSE}
ParsePrimers <- function(fastaFile) {
  fastaString <- paste(readLines(fastaFile), collapse = "\n")
  fastaList <- strsplit(fastaString, split = ">")
  fastaList <- fastaList[[1]][fastaList[[1]] != ""]
  newLineSplit <- function(inputLine) {
    sequenceList <- strsplit(inputLine, split = "\n")[[1]]
    nucSequence <- sequenceList[2:length(sequenceList)]
    nucSequence <- paste(nucSequence, collapse = "")
    sequenceList <- c(sequenceList[1], strsplit(nucSequence, "\\.\\.\\."))
    sequenceDF <- data.frame(Primer = c(sequenceList[[2]][1], sequenceList[[2]][2]))
    colnames(sequenceDF) <- sequenceList[[1]]
    return(sequenceDF)
  }
  fastaDFList <- lapply(fastaList, newLineSplit)
  fastaDF <- do.call("cbind", fastaDFList)
  return(fastaDF)
}

primerDF <- ParsePrimers(fastaFile = primerPath)
```

# Load files:
```{r pressure, echo=FALSE}
NameShortener <- function(name) {
  # name <- substr(name, 1, 20)
  name <- gsub("_[^_]+$", "", name)
  return(name)
}

FastqLoader <- function(fileName, pathToData = filteredFolder) {
  #namePattern <- paste0(fileName, "*")
  loadedFile <- readFastq(dirPath = pathToData, pattern = fileName)
  print(paste("Searching for", fileName))
  print(paste("Found", length(list.files(pathToData, pattern = fileName)), "files."))
  return(c(loadedFile, fileName))
}

# Identify all files:
filteredFolder <- "~/Documents/bearNGSGenotype/Filtered_data2"


# Scat half plate
pfun <- function(number) {
    paste0("S",LETTERS[1:8], number)
}

# Scat half plate
prover <- as.vector(sapply(7:12, pfun))
which(prover=="SE12")
prover <- prover[-c(2,3,4,31,42,45)]
#prover <- prover[-19]

# Scat full plate
filteredFolder <- "~/Documents/bearNGSGenotype/Filtered_data"
pfun <- function(number) {
    paste0(LETTERS[1:8], number)
}

#Scat full plate
prover <- as.vector(sapply(1:12, pfun))
which(prover=="H12")
prover <- prover[-c(4, 12, 19, 20, 23, 25, 26, 28, 33, 43, 46, 50, 52, 73, 79, 96)]

resToCheck2 <- data.frame()

prover <- c("A1", "SA11")

for(i in prover) {
    
    i <- prover[1]
    inputFiles <- list.files(filteredFolder, pattern = paste0("_", i, "B"))
    inputFiles2 <- unlist(lapply(inputFiles, NameShortener))
    pdfname <- paste0(unlist(strsplit(inputFiles2[1], "_"))[2], ".pdf")
    par(mfrow=c(4,3))
    loadedFastqFiles <- lapply(inputFiles, FastqLoader)
    gentypRes <- PlotAllSamples(loadedFiles = loadedFastqFiles, 
                                writeTo = pdfname, 
                                primerInfo = primerDF)
    
    resToCheck2 <- rbind(resToCheck2, reformat(gentypRes))
}



inputFiles <- list.files(filteredFolder, pattern = "_A3_1.fastq.gz")
inputFiles <- unlist(lapply(inputFiles, NameShortener))
#inputFiles <- unique(inputFiles)

# Load all files:
loadedFastqFiles <- lapply(inputFiles, FastqLoader)
```

# Identify genotypes and plot peaks:
```{r pressure, echo=FALSE}
FragLength <- function(sample, primerF = "AAAAA", primerR = "AAAAA", range = 0:100) {
  lengthF <- nchar(primerF)
  lengthR <- nchar(primerR)
  fragl <- ShortRead::width(ShortRead::sread(sample))
  lengthTot <- lengthF + lengthR + fragl
  lengthTot <- lengthTot[lengthTot > min(range) & lengthTot < max(range)]
  tabl <- table(lengthTot)
  if(length(tabl) > 8) {
    tabl <- sort(tabl, decreasing = TRUE)[1:8]
  }
  # tabl <- as.table(tabl[tabl > cutoff * sum(tabl)])
  names(attributes(tabl)$dimnames) <- "AlleleLength"
  tabl <- tabl[order(as.numeric(names(tabl)))]
  return(tabl)
}

MostLikely <- function(results) {
  AlleleCandidate1 <- results[which.max(names(results))]
  resultsShort <- results[results != AlleleCandidate1]
  AlleleCandidate2 <- resultsShort[which.max(names(resultsShort))]
  results_sorted <- sort(results)
  max_peak <- results_sorted[length(results_sorted) - 0:1]
  AlleleCandidate3 <- max_peak[1]
  AlleleCandidate4 <- max_peak[2]
  AC <- list(AC1 = AlleleCandidate1[1], AC2 = AlleleCandidate2, 
             AC3 = AlleleCandidate3, AC4 = AlleleCandidate4)
  GT <- c(AC$AC1, AC$AC3)
  names(rev(GT))
}

BarCol <- function(Allelelength, highlight) {
  c1 <- rep(wesanderson::wes_palette("Chevalier1")[3], length(Allelelength))
  c1[which(names(Allelelength) %in% highlight)] <- wesanderson::wes_palette("Chevalier1")[2]
  c1
}

PlotSample <- function(sampleName, sampleTab, sampleML) {
  barplot(sampleTab,
          col = BarCol(sampleTab, highlight = sampleML),
          main = sampleName,
          xlab = paste(sampleML, collapse = "/"))
}




PlotAllSamples <- function(loadedFiles, writeTo, primerInfo) {
  pdf(writeTo)
  outputDF <- data.frame(Allele = character(), LengthShort = numeric(), LengthLong = numeric())
  for(i in loadedFiles) {
      dfRow <- nrow(outputDF) + 1
      if(length(i[[1]]) == 0) {
          barplot(c(0, 0),
                  main = unlist(i[2]),
                  xlab = "0/0")
          outputDF[dfRow, 1] <- unlist(i[2])
          outputDF[dfRow, 2:3] <- c(NA,NA)
    } else {
      primerName <- unlist(strsplit(unlist(i[2]), "_"))[1]
      primerFR <- primerInfo[, colnames(primerInfo) == primerName]
      tryCatch(
        {
        tabSample <- FragLength(sample = i[[1]], range = 10:100)
      mlSample <- MostLikely(tabSample)
      PlotSample(sampleName = unlist(i)[2], sampleTab = tabSample, sampleML = mlSample)
      outputDF[dfRow,1] <- unlist(i[2])
      outputDF[dfRow, 2:3] <- as.numeric(mlSample)
        },
        error=function(e) {
            message('An Error Occurred')
            print(e)
            outputDF[dfRow,1] <- unlist(i[2])
            outputDF[dfRow, 2:3] <- c(NA,NA)
        },
        #if a warning occurs, tell me the warning
        warning=function(w) {
            message('A Warning Occurred')
            print(w)
            return(NA)
        }
      )
      
#      tabSample <- FragLength(sample = i[[1]], range = 10:100)
#      mlSample <- MostLikely(tabSample)
#      PlotSample(sampleName = unlist(i)[2], sampleTab = tabSample, sampleML = mlSample)
#      outputDF[dfRow,1] <- unlist(i[2])
#      outputDF[dfRow, 2:3] <- as.numeric(mlSample)
      #outputDF <- rbind(col.names(outputDF), )
    }
  }
  dev.off()
  return(outputDF)
}

gentypRes <- PlotAllSamples(loadedFiles = loadedFastqFiles, writeTo = "test.pdf", primerInfo = primerDF)

reformat <- function(alleles) {
    gentypResTrans <- t(alleles[,-1])
    colnames(gentypResTrans) <- alleles$Allele
    sample <- colnames(gentypResTrans)[1]
    sample <- unlist(strsplit(sample, "_"))[2]
    resdf <- data.frame(G10B_1   = gentypResTrans[1,1],
                        G10B_2   = gentypResTrans[2,1],
                        G10L_1   = gentypResTrans[1,2],
                        G10L_2   = gentypResTrans[2,2],
                        G1A_1    = gentypResTrans[1,3],
                        G1A_2    = gentypResTrans[2,3],
                        G1D_1    = gentypResTrans[1,4],
                        G1D_2    = gentypResTrans[2,4],
                        MU05_1   = gentypResTrans[1,5],
                        MU05_2   = gentypResTrans[2,5],
                        MU09_1   = gentypResTrans[1,6],
                        MU09_2   = gentypResTrans[2,6],
                        MU10_1   = gentypResTrans[1,7],
                        MU10_2   = gentypResTrans[2,7],
                        MU15_1   = gentypResTrans[1,8],
                        MU15_2   = gentypResTrans[2,8],
                        MU23_1   = gentypResTrans[1,9],
                        MU23_2   = gentypResTrans[2,9],
                        MU50_1   = gentypResTrans[1,10],
                        MU50_2   = gentypResTrans[2,10],
                        MU59_1   = gentypResTrans[1,11],
                        MU59_2   = gentypResTrans[2,11])
    row.names(resdf) <- sample
    return(resdf)
}
 
reformat(gentypRes)

# sortera bort inkorrekta sekvenser av samma längd, längre distans än sekvensfel
```
