---
title: "ngsgenotype_snp"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Detect and load files:
```{r}
# Load libraries:
library(ShortRead)
library(wesanderson)
source("genotyping_functions.r")

# List files:
snpCsvFiles <- list.files("Genotyped_SNPs/", pattern = ".csv", full.names = TRUE)

# Load csv files into a list of dataframes:
loadedFiles <- lapply(snpCsvFiles, read.table, sep = ',', header = TRUE, check.names = FALSE)

# Row bind all separate dataframes into one:
allGenotypes <- do.call(rbind, loadedFiles)

# Replace NA values:
allGenotypes[is.na(allGenotypes)] <- ''
```

# Filter samples and loci:
```{r}
GenotypeFiltering <- function(inGenotypes, minPresent = 75, maxPresent = 10000, createRDS = FALSE, rdsName = "filteredGenotypes.rds") {
  nLoci <- ncol(inGenotypes) - 1
  sumEmpties <- function(x) {return(sum(grepl("^$", x)))}
  filteredGenotypes <- inGenotypes[apply(inGenotypes, 1, sumEmpties) <= (nLoci - minPresent), ]
  filteredGenotypes <- filteredGenotypes[apply(filteredGenotypes, 1, sumEmpties) >= (nLoci - maxPresent), ]
  if(createRDS == TRUE) {
    saveRDS(filteredGenotypes, file = rdsName)
  }
  return(filteredGenotypes)
}

# Remove controls:
sampleGenotypes <- allGenotypes[!grepl("NEGATIVE|POSITIVE", allGenotypes$Sample), ]

# Filter out sex loci:
sexString <- "Ua-snp-01y|Ua-snp-02y|Ua-snp-03y|Ua-snp-04y|Ua-snp-196-x|Ua-snp-248-x|Ua-snp-236-x"
autoGenotypes <- sampleGenotypes[, !grepl(sexString, colnames(sampleGenotypes))]
saveRDS(autoGenotypes, file = "autoGenotypes.rds")
sexGenotypes <- cbind(Sample = sampleGenotypes$Sample, sampleGenotypes[, grepl(sexString, colnames(sampleGenotypes))])
saveRDS(sexGenotypes, file = "sexGenotypes.rds")

# Filter away meat samples:
scatGenotypes <- autoGenotypes[!grepl("^M|^VLT", autoGenotypes$Sample),]

# Filter out samples with minimum 75/67 working loci as well as the rejected ones:
filteredGenotypes75 <- GenotypeFiltering(scatGenotypes, minPresent = 75, createRDS = TRUE, rdsName = "filteredGenotypes75.rds")
filteredGenotypes67 <- GenotypeFiltering(scatGenotypes, minPresent = 67, maxPresent = 74, createRDS = FALSE)
rejectedGenotypes <- GenotypeFiltering(scatGenotypes, minPresent = 0, maxPresent = 66, createRDS = TRUE, rdsName = "rejectedGenotypes.rds")

# Filter out empty loci:
presentGenotypes <- filteredGenotypes75[, !lapply(filteredGenotypes75, unique) == ''] # does this matter with allelematch?
```

# Collapse samples to individuals:
```{r}
library(allelematch)

SplitLocus <- function(x, dataSet) {
  columnName <- colnames(dataSet)[x]
  print(columnName)
  twoCols <- do.call(rbind, strsplit(dataSet[, x], ""))
  colnames(twoCols) <- c(paste0(columnName, "1"), paste0(columnName, "2"))
  return(twoCols)
}

# Split loci into two columns:
splittingGTs <- presentGenotypes
splittingGTs[splittingGTs == ''] <- "NN" # change missing data to NN
splitDataset <- cbind(Sample = splittingGTs$Sample, do.call(cbind, lapply(1:ncol(splittingGTs[, -c(1)]), SplitLocus, splittingGTs[, -c(1)])))

# Load data:
snpLoadedSplit <- allelematch::amDataset(multilocusDataset = splitDataset, missingCode = "N", indexColumn = "Sample")
bearAlleleKey <- rep(1:(ncol(splitDataset[, -c(1)]) / 2), each = 2)

# Find optimal mismatch parameters:
amProfileSplit <- allelematch::amUniqueProfile(amDatasetFocal = snpLoadedSplit, multilocusMap = bearAlleleKey, doPlot = TRUE, alleleMismatch = c(0:45)) # 13 is the best
uniqueBears <- allelematch::amUnique(snpLoadedSplit, multilocusMap = bearAlleleKey, alleleMismatch = 13) # change 13 to what amUniqueProfile picks
allelematch::summary.amUnique(object = uniqueBears, html = "unique_bears_75_13.html") # generate report
```














