---
title: "ngsgenotype_snp"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Detect and load files:
```{r}
# Load libraries:
library(ShortRead)
library(wesanderson)
source("genotyping_functions.r")

# List files:
snpCsvFiles <- list.files("Genotyped_SNPs/", pattern = ".csv", full.names = TRUE)

# Load csv files into a list of dataframes:
loadedFiles <- lapply(snpCsvFiles, read.table, sep = ',', header = TRUE, check.names = FALSE)

# Row bind all separate dataframes into one:
allGenotypes <- do.call(rbind, loadedFiles)

# Replace NA values:
allGenotypes[is.na(allGenotypes)] <- ''
```

# Filter samples and loci:
```{r}
GenotypeFiltering <- function(inGenotypes, minPresent = 75, maxPresent = 10000, createRDS = FALSE, rdsName = "filteredGenotypes.rds") {
  nLoci <- ncol(inGenotypes) - 1
  sumEmpties <- function(x) {return(sum(grepl("^$", x)))}
  filteredGenotypes <- inGenotypes[apply(inGenotypes, 1, sumEmpties) <= (nLoci - minPresent), ]
  filteredGenotypes <- filteredGenotypes[apply(filteredGenotypes, 1, sumEmpties) >= (nLoci - maxPresent), ]
  if(createRDS == TRUE) {
    saveRDS(filteredGenotypes, file = rdsName)
  }
  return(filteredGenotypes)
}

# Remove controls:
sampleGenotypes <- allGenotypes[!grepl("NEGATIVE|POSITIVE", allGenotypes$Sample), ]

# Filter out sex loci:
sexString <- "Ua-snp-01y|Ua-snp-02y|Ua-snp-03y|Ua-snp-04y|Ua-snp-196-x|Ua-snp-248-x|Ua-snp-236-x"
autoGenotypes <- sampleGenotypes[, !grepl(sexString, colnames(sampleGenotypes))]
saveRDS(autoGenotypes, file = "autoGenotypes.rds")
sexGenotypes <- cbind(Sample = sampleGenotypes$Sample, sampleGenotypes[, grepl(sexString, colnames(sampleGenotypes))])
saveRDS(sexGenotypes, file = "sexGenotypes.rds")

# Filter away meat samples:
scatGenotypes <- autoGenotypes[!grepl("^M|^VLT", autoGenotypes$Sample),]

# Filter out samples with minimum 75/67 working loci as well as the rejected ones:
filteredGenotypes75 <- GenotypeFiltering(scatGenotypes, minPresent = 75, createRDS = TRUE, rdsName = "filteredGenotypes75.rds")
filteredGenotypes67 <- GenotypeFiltering(scatGenotypes, minPresent = 67, maxPresent = 74, createRDS = FALSE)
rejectedGenotypes <- GenotypeFiltering(scatGenotypes, minPresent = 0, maxPresent = 66, createRDS = TRUE, rdsName = "rejectedGenotypes.rds")

# Filter out empty loci:
presentGenotypes <- filteredGenotypes75[, !lapply(filteredGenotypes75, unique) == ''] # does this matter with allelematch?
```

# Collapse samples to individuals:
```{r}
# Load libraries:
library(poppr)
library(adegenet)

# Load data:
genotypeGenind <- adegenet::df2genind(presentGenotypes[, -c(1)], ploidy = 2, ind.names = presentGenotypes$Sample, sep = "")
genotypeGenind
genotypeGencln <- poppr::as.genclone(genotypeGenind)
genotypeGencln

# Collapse genotypes:
mll(genotypeGencln) <- "original" # reset genotypes
system.time(calculatedDistances <- BearDist(presentGenotypes)) # calculate BearDists (TM)
saveRDS(calculatedDistances, file = "calculatedDistances.rds") # save calculated distances
calculatedDistances # print BearDists (TM)
hist(calculatedDistances) # histogram of BearDists (TM)
threshold <- (ncol(presentGenotypes) - 1) - 67 # make a threshold where 67 loci needs to be similar between samples
mlg.filter(genotypeGencln, distance = calculatedDistances) <- threshold # collapse based on threshold, using BearDists (TM)
genotypeGencln
mll(genotypeGencln)

# Assign genotypes:
gtAndSnps <- as.data.frame(cbind(genotypes = genotypeGencln@mlg@mlg$contracted, genotypeGencln$tab)) # add genotypes to SNP table
uniqueSamples <- rownames(gtAndSnps)[match(unique(gtAndSnps$genotypes), gtAndSnps$genotypes)] # take first sample of each genotype
# Pull those samples from the sample sheets

# Mockup:
# 1. add a column to the "database" with ind-names
# 2. load the "database" (without ind-names) and new samples together
# 3. collapse individuals (will collapse the new ones to an existing genotype)
# 4. check if collapsed genotypes contain a sample with an existing ind-name, if so add new ones to database under that name
# 5. if individuals do not exist, add them to database with new names.
```














