---
title: "ngsgenotype_snp"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries and settings:
```{r}
library(ShortRead)
library(wesanderson)

filteredFolder <- "./Filtered_data_minion/SNP_filtered/"
filteredFolder <- "./Filtered_data_BMK/SNP_filtered/"
```

# Detect files:
```{r}
inputFiles <- list.files(filteredFolder, pattern = ".fastq.gz")
```

# Load files:
```{r}
FastqLoader <- function(fileName, pathToData = filteredFolder) {
  loadedFile <- ShortRead::readFastq(dirPath = pathToData, pattern = fileName)
  print(paste("Searching for", fileName))
  print(paste("Found", length(list.files(pathToData, pattern = fileName)), "files."))
  return(c(loadedFile, fileName))
}

PickGenotype <- function(srVector, cutoff) {
  srObject <- srVector[[1]]
  srTable <- ShortRead::tables(srObject)$top
  if(length(srTable) > 1) {
    topSNP <- srTable[1]
    sndSNP <- srTable[2]
  } else if(length(srTable) == 1) {
    topSNP <- srTable[1]
    sndSNP <- srTable[1]
  } else {
    topSNP <- 0
    sndSNP <- 0
  }
  if((topSNP + sndSNP) > cutoff[1]) {
    if(((sndSNP / topSNP) * 100) > cutoff[2]) {
      return(c(srVector[[2]], paste0(names(topSNP), names(sndSNP))))
    } else {
      # return(c(srVector[[2]], names(topSNP), names(topSNP)))
      return(c(srVector[[2]], paste0(names(topSNP), names(topSNP))))
    }
  } else {
    # return(c(srVector[[2]], "-", "-"))
    return(c(srVector[[2]], NA))
  }
}

RunLocus <- function(locusName, allFiles, filteredFolder, locusCutoffs = c()) {
  print(paste0("Running locus: ", locusName))
  locusFiles <- allFiles[grepl(locusName, inputFiles)]
  loadedFastqFiles <- lapply(locusFiles, FastqLoader, filteredFolder)
  if(locusName %in% names(locusCutoffs)) { # if a specific cutoff is set for this locus
    cutoff = locusCutoffs[locusName][[1]]
  } else { # else use standard settings
    cutoff = c(10, 1) # minimum 10 reads, minimum 10% to be a heterozygote
  }
  genotypeRes <- lapply(loadedFastqFiles, PickGenotype, cutoff)
  rm(loadedFastqFiles) # remove to conserve memory
  gc() # run garbage collection to free memory
  genotypeCols <- do.call(rbind, genotypeRes)
  genotypeCols <- as.data.frame(genotypeCols)
  # colnames(genotypeCols) <- c("Sample", paste0(locusName, "_1"), paste0(locusName, "_2"))
  colnames(genotypeCols) <- c("Sample", locusName)
  genotypeCols$Sample <- gsub(".+_|BMK.+", "", genotypeCols$Sample)
  return(genotypeCols)
}

RunAllLoci <- function(inputFiles, filteredFolder, locusCutoffs = c()) {
  lociNames <- unique(gsub("_.+", "", inputFiles)) # remove everything after and including the first underscore
  genotypesList <- lapply(lociNames, RunLocus, inputFiles, filteredFolder, locusCutoffs)
  allGenotypes <- Reduce(function(x, y) merge(x, y, by = "Sample"), genotypesList)
  return(allGenotypes)
}

# Specify if any loci should have specific cutoffs:
locusCutoffs <- c()
locusCutoffs <- rep(list(c(10, 10)), length(unique(gsub("_.+", "", inputFiles))))
names(locusCutoffs) <- unique(gsub("_.+", "", inputFiles))
locusCutoffs
# locusCutoffs$`locus name` <- c(10, 10) # add 'locus name' and its designated cutoffs

# Load all samples and calculate all loci:
allGenotypes <- RunAllLoci(inputFiles, filteredFolder, locusCutoffs)

# If genotypes are returned as only one letter it is because of the fact that there are empty reads in the fastq
```

# Testing:
```{r}
loadedFile <- ShortRead::readFastq(dirPath = filteredFolder, pattern = "Ua-snp-146_minion_r4_merged_barcode203.fastq.gz")
loadedFile <- ShortRead::readFastq(dirPath = filteredFolder, pattern = "Ua-snp-125_minion_r4_merged_barcode208.fastq.gz")
View(tables(loadedFile)$top[1:2])
View(tables(loadedFile)$top)
tables(loadedFile)$top

PickGenotype(c(loadedFile, "Ua-snp-146_minion_r4_merged_barcode203.fastq.gz"), c(100, 10))
locusGenotypes <- RunLocus("Ua-snp-146", inputFiles, filteredFolder, locusCutoffs)
```

# Testing popgen:
```{r}
library(poppr)
library(snpReady)

# head(allGenotypes)
# allGenotypesEdit <- allGenotypes[, -c(1)]
# rownames(allGenotypesEdit) <- allGenotypes$Sample
# loadedGenotypes <- snpReady::raw.data(data = as.matrix(allGenotypesEdit), frame = "wide", base = TRUE, sweep.sample = 0.5, call.rate = 0.95, maf = 0.10, imput = FALSE)

library(adegenet)
# Genind path:
genotypeGenind <- adegenet::df2genind(allGenotypes[, -c(1)], ploidy = 2, ind.names = allGenotypes$Sample, sep = "")
genotypeGenind
genotypeGencln <- poppr::as.genclone(genotypeGenind)
# mll(genotypeGencln) <- "original"
# mll(genotypeGencln) <- "contracted"
# mlg.filter(genotypeGencln, distance = diss.dist)
# genotypeGencln
# poppr::mlg(genotypeGencln)
# library(pegas)
# genotypeLoci <- pegas::genind2loci(genotypeGenind)
# genotypeLoci
hist(diss.dist(genotypeGenind, percent = FALSE, mat = FALSE))
# hist(bitwise.dist(genotypeGencln, percent = FALSE, mat = FALSE)) # gives exactly the same plot

# mlg.filter(genotypeGencln, distance = bitwise.dist)
# mlg.filter(genotypeGencln, distance = diss.dist) <- .Machine$double.eps^0.5
mlg.filter(genotypeGencln, distance = diss.dist) <- 30 # set a high threshold based on histogram
genotypeGencln
mll(genotypeGencln)

# Genlight path:
# allGenotypesEdit <- allGenotypes[, -c(1)]
# allGenotypesEdit <- allGenotypesEdit[, -c(6, 29)]
# rownames(allGenotypesEdit) <- allGenotypes$Sample
# genGenlight <- adegenet::as.genlight(allGenotypesEdit)
# genGenlight
# genotypeSnpclone <- poppr::as.snpclone(genGenlight)
```

# Initial steps to collapsing individuals:
```{r}
library(poppr)
library(adegenet)

genotypeGenind <- adegenet::df2genind(allGenotypes[, -c(1)], ploidy = 2, ind.names = allGenotypes$Sample, sep = "")
genotypeGenind
genotypeGencln <- poppr::as.genclone(genotypeGenind)

mll(genotypeGencln) <- "original"
# genotypeGencln
# mll(genotypeGencln)
hist(diss.dist(genotypeGenind, percent = FALSE, mat = FALSE))
mlg.filter(genotypeGencln, distance = diss.dist) <- 30 # set a high threshold based on histogram
genotypeGencln
mll(genotypeGencln)

# assign individual names as populations for continuous SNP-database, and add samples to database under assigned population?
```

